pkg/virt-controller/watch/vmi.go

```diff
@@ -870,6 +871,22 @@ func prepareVMIPatch(oldVMI, newVMI *virtv1.VirtualMachineInstance) ([]byte, err
 		log.Log.V(3).Object(oldVMI).Infof("Patching VMI activePods")
 	}
 
+	if !equality.Semantic.DeepEqual(newVMI.Status.Phase, oldVMI.Status.Phase) {
```

Comment:
This should be cover by
```
// If we detect a change on the vmi we update the vmi
	vmiChanged := !equality.Semantic.DeepEqual(vmi.Status, vmiCopy.Status) || !equality.Semantic.DeepEqual(vmi.Finalizers, vmiCopy.Finalizers) || !equality.Semantic.DeepEqual(vmi.Annotations, vmiCopy.Annotations) || !equality.Semantic.DeepEqual(vmi.Labels, vmiCopy.Labels)
	if vmiChanged {
		key := controller.VirtualMachineInstanceKey(vmi)
		c.vmiExpectations.SetExpectations(key, 1, 0)
		_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Update(context.Background(), vmiCopy)
		if err != nil {
			c.vmiExpectations.LowerExpectations(key, 1, 0)
			return err
		}
	}

	return nil
```

Reply:
As far as I was able to test, it is not. You're completely right that this code should update the VMi Phase. But for some reason, when the node gets down, it is not updating the VMi Phase.

Reply:
So this sounds like a bug to me. Could you have a look at this?

Reply:
I've taken a look at it and the update part is not executed until the VMi is already in `Failed` state. 

Looking at this [line](https://github.com/jcanocan/kubevirt/blob/56012c131325f2bb24714bf69ea40beb35624418/pkg/virt-controller/watch/vmi.go#L752) :

```go
// VMI is owned by virt-handler, so patch instead of update
if vmi.IsRunning() || vmi.IsScheduled() {
		patchBytes, err := prepareVMIPatch(vmi, vmiCopy)
		if err != nil {
			return fmt.Errorf("error preparing VMI patch: %v", err)
		}

		if len(patchBytes) > 0 {

			_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, []byte(patchBytes), &v1.PatchOptions{})
			// We could not retry if the "test" fails but we have no sane way to detect that right now: https://github.com/kubernetes/kubernetes/issues/68202 for details
			// So just retry like with any other errors
			if err != nil {
				return fmt.Errorf("patching of vmi conditions and activePods failed: %v", err)
			}
		}

		return nil
	}

```
it makes sense since the VMi is on `Running` state. Therefore, it will execute the patch section of code and `return nil`, so the update is not executed until the VMi is on a `Phase` different from `Running` or ` Scheduled`. IMHO, this is right, and we should keep the `Phase` update in the patch section.

If we don't patch the `Phase` in this point, the launcher_respawn time is the same as the baseline case (up to 100s).

Reply:
As long as the VMI is considered running/scheduled, `virt-controller` will only patch it. As @jcanocan pointed out, therefore Update is never called. See L767.

Reply:
You are both correct, sorry for the confusion.

Could we change this slightly:

```
if newVMI.Status.Phase == v1.Failed &&  oldVMI.Status.Phase == v1.Running  && pod == nil{
    patchOps = append(patchOps,  `{ "op": "test", "path": "/status/phase", "value": "Running" }`)
    patchOps = append(patchOps, `{ "op": "replace", "path": "/status/phase", "value": "Failed" }`)
}
```
Note: we need to pass the pod into the function. 
Note 2: no need for the `Marshal`

The reason behind this is to limit when we make this patch as much as we can. Right now we have a clear separation of when the handler or controller acts and I want to make sure this is seen as an exception. Also comment would be really good to have here.

Reply:
Done.